L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+  string country = 1;
+}
+
+message CountryByUsernameRequest {
+  string username = 1;
+}
+
+message CountryByUsernameResponse {
+  string country = 1;
+}
+```
+
+```cpp
+UserInfoRequest MakeUserInfoRequest(const std::string& userid) {
+  UserInfoRequest request;
+  request.set_userid(userid);
+  return request;
+}
+
+class TestServiceImpl : public TestService::ReactiveService {
+ public:
+  /// Request handler that makes one upstream request
+  AnyPublisher<CountryByUserIdResponse> CountryByUseridHandler(
+      CountryByUseridRequest request) override {
+    return AnyPublisher<CountryByUserIdResponse>(Pipe(
+        // Make RPC call to get userinfo
+        upstream_service_.GetUserInfo(
+            MakeUserInfoRequest(request.userid())),
+        // Convert the result to a response of the proper type
+        Map([](UserInfoResponse response) {
+          CountryByUseridResponse final_response;
+          final_response.set_country(response.country());
+          return final_response;
+        })));
+  }
+
+  /// Request handler that makes two upstream requests in a row
+  AnyPublisher<CountryByUsernameResponse> CountryByUsernameHandler(
+      CountryByUsernameRequest request) override {
+    ResolveUsernameRequest resolve_username_request;
+    resolve_username_request.set_username(request.username());
+
+    return AnyPublisher<CountryByUsernameResponse>(Pipe(
+        // First, resolve the username to a user id
+        upstream_service_.ResolveUsername(resolve_username_request),
+        // Then resolve the user id to userinfo
+        ConcatMap([this](ResolveUsernameResponse response) {
+          return upstream_service_.GetUserInfo(
+              MakeUserInfoRequest(response.userid()));
+        }),
+        // Then take the resulting userinfo and construct a response
+        Map([](UserInfoResponse response) {
+          CountryByUsernameResponse final_response;
+          final_response.set_country(response.country());
+          return final_response;
+        })));
+  }
+
+ private:
+  // This would have to be initialized of course. Omitted here.
+  UpstreamService::ReactiveStub upstream_service_;
+};
+```
+
+*Notes:*
+
+* `Pipe` is a function in rs that lets you "pipe" a Publisher through a series of operators. The first argument is a Publisher/stream, subsequent arguments are operators on that stream.
+* `Pipe(a, b, c)` is essentially the same as `c(b(a))`. Having `Pipe` like this isn't necessary. Another option is to let the operators be methods on a base Publisher type like RxJava does, but then it's impossible for user code to add their own first-class operators.
+* `ConcatMap` is similar to `Map` but it lets the mapper function return a Publisher instead of a value directly. I would personally call it `FlatMap` but I've stuck to ReactiveX's naming (there is a `FlatMap` in ReactiveX, but it behaves in an unintuitive way and its implementation requires an unbounded buffer).
+
+
+#### Client streaming: Sum
```

Comment:
Please provide an example whereby a client makes calls to multiple backends, with variants:
1. waits for all responses to come back, and then forms a coalesced response
2. waits for one response to come back, forwards it, and cancels all other requests

These patterns are incredibly common in real systems, and providing good support for them needs to be a requirement for any API addition.

Reply:
I will update with examples demonstrating these use cases soon.

Reply:
I have added those examples now. For 1, see `RenderShoppingCart`. For 2, see `CountryByUseridRace`.

Notes:

If you don't like `Zip` returning a tuple and the `Splat` helper, `Zip` could be changed so that its last parameter is a function with one parameter per input stream that gets to combine the values (RxJava does this). Personally I can't decide which I prefer.

rs does not currently have variants of `Zip` and `Merge` that take a list of input streams; they are simply variadic functions. This is not a fundamental limitation, I just haven't written them yet.

It's a very contrived example, but you might be interested in knowing that the pattern of making one request twice and letting them race can be implemented as an operator (I'm cheating with C++14 here but it can be done in C++11 as well with `AnyPublisher` (with some runtime overhead) or more verbosity):

```cpp
template <typename Input>
auto TwinRace(const Input &input) {
  return Pipe(
      Merge(input, input),
      Take(1));
}
```

which could then be used as an operator like any other. With the addition of a sleep/delay operator (not in the current implementation), you could do optimistic retries:

```cpp
template <typename Input>
auto TwinRace(const Input &input) {
  using namespace std::chrono_literals;
  return Pipe(
      Merge(input, Delay(100ms, input)),
      Take(1));
}
```

Perhaps even better than `Take(1)` would be to use the [race operator](http://rxmarbles.com/#race) but rs doesn't have that currently.

---

With this I consider this issue resolved, let me know if you disagree.